stat/spatial/spatial.go

```diff
@@ -43,11 +42,21 @@ func GetisOrdGStar(i int, data, weights []float64, locality mat.Matrix) float64
 	n := float64(len(data))
 	mean, std := stat.MeanStdDev(data, weights)
 	var dwd, dww, sw float64
-	for j, v := range data {
-		w := locality.At(i, j)
-		sw += w
-		dwd += w * v
-		dww += w * w
+	if band, ok := locality.(mat.Banded); ok {
```

Comment:
Sorry, I don't have time to do a full review at the moment, but can we abstract this somehow with an interface? Something like "non-zero apply"?  I think we'd like to discourage this kind of type switching outside the matrix package, and instead should create good interfaces to allow higher packages to be agnostic.